### PR TITLE
Feature/request body validation

### PR DESCRIPTION
**Pull request by vaibhavatul47**
### Summary

1. Adding functionality to validate request body in hmac-auth. This is optional and can be configured when enabling the Hmac-Auth plugin for an API.

2. Making `date` or `x-date` header mandatory in Hmac-headers.

#### 1. Request Body validation
Currently, Hmac-Auth plugin allows validation of only request headers. There are requirements in real life projects when the validity of request body is also to be verified. To solve this problem I have introduced the usage of `digest` header. This `digest` header will contain the Base64 encoded Sha256 hash of the request body.

If an API has `hmac-auth` plugin enabled and `validate-request-body` config is enabled then for every request of this API received at KONG, Hmac-Auth plugin will also validate the integrity of request body, needless to say, this check will be performed only if request contains any body

If this config( `validate-request-body`) is enabled on an API then every request **must** contain `digest` header inside `hmac-headers` parameter.

#### 2. Mandatory header values in Hmac-headers parameter
Currently, although the plugin validates clock-skew but it is no guarantee that the `date` or `x-date` header used in clock skew validation is not tampered.
Also, according to  [ietf draft](https://tools.ietf.org/html/draft-cavage-http-signatures-06#section-2.3) it is madatory to include `date` headers in Hmac-header parameter.
If `validate-request-body` config is enabled for an API then Hmac-header should also must contain `digest` header.

### Tests added
* should pass with GET when body validation enabled
* should pass with POST when body validation enabled and digest header present
* should not pass with GET with date header missing from hmac-headers
* should not pass with POST when body validation enabled and digest header missing
* should not pass with POST when body validation enabled and digest header missing from hmac-headers
* should not pass with POST when body validation enabled and postBody is tampered
* should not pass with POST when body validation enabled and digest header is tampered

### CODE REVIEW

**Comment by thibaultcha** [comment]
Hi,

Thanks for the contribution, but please rebase your changes on next to keep the history clean of unrelated changes. Additionally, if this is to be considered a default behavior as you intended, I believe you are missing migrations to enable this flag in already existing configurations.

File: kong.iml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (p0pr0ck5) [inline kong.iml]: seems this file shouldn't be part of this PR?

File: kong/plugins/hmac-auth/access.lua [hunk 2]
```diff
@@ -79,30 +84,31 @@ local function create_hash(request, hmac_params, headers)
       signing_string = signing_string.."\n"
     end
   end
+
   return ngx_sha1(hmac_params.secret, signing_string)
 end
 
-local function is_digest_equal(digest_1, digest_2)
+local function is_signature_equal(signature_1, signature_2)
```
Comment (p0pr0ck5) [inline kong/plugins/hmac-auth/access.lua]: what are we changing this function name and variables? this dirties the commit history with no real value, and decreases the atomicity of this commit.

File: kong.iml (hunk 1 above)
Comment (vaibhavatul47) [inline kong.iml]: Yes, committed this file by mistake. I'll remove it.

File: kong/plugins/hmac-auth/access.lua (hunk 2 above)
Comment (vaibhavatul47) [inline kong/plugins/hmac-auth/access.lua]: After this change, `digest` has a separate role: To signify Sha256 hash of request body. I renamed this function and its variable to avoid any confusion in future. Also, imo the function name should sync with its functionality, it is equating signatures so its name should justify this.

**Comment by vaibhavatul47** [comment]
Hi all, I've made multiple commits in this PR. As per development instructions, all changes should be in a single commit, should open a new PR with all my changes in one commit?

**Comment by vaibhavatul47** [comment]
I see here **travis-ci** build failed for one job, whereas same travis-ci build completed successfully on my repo on same branch. I have verified both used same config. What could be the reason for different build results for same code?
Below are screenshots of travis-ci builds:
Build result of Mashape-Kong repo
screenshot

Build result of my forked repo:
screenshot

**Comment by p0pr0ck5** [comment]
@vaibhavatul47 you can make multiple commits on this PR if you like, and they can be squashed in the merge. Alternatively, you can commit your changes with `--amend`, and then `git push -f` to your branch for the changes to the PR to take affect. You still need to rebase though, as this merge request brings in a lot of unrelated commits.

As for the failed tests, it is unrelated to your PR; please ignore that for now.

**Comment by p0pr0ck5** [comment]
Actually, it seems like the two points you've noted in the summary (request body validation and mandatory date/x-date headers) are unrelated. To maintain atomicity, would you mind splitting these into two separate PRs, one for request body validation, and one for mandatory date headers?

**Comment by vaibhavatul47** [comment]
Splitting into two different PRs one for each change looks good approach. On it.
Will update once ready.

**Comment by vaibhavatul47** [comment]
New PRs:
https://github.com/Mashape/kong/pull/2418
https://github.com/Mashape/kong/pull/2419

**Comment by p0pr0ck5** [comment]
Thanks, will close this one.
